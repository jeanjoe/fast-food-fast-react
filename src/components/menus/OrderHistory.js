import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { getOrderHistory } from "../../actions/menuActions";
import { Container, Item, Breadcrumb } from "semantic-ui-react";
import Alert from "../notifications/Alert";
import OrderHistoryItem from "./OrderHistoryItem";

export class OrderHistory extends Component {
  constructor(props) {
    super(props);
    this.state = {
      loading: true,
      orderHistory: [],
      orderHistoryError: null
    };
  }

  componentDidMount() {
    const { getOrderHistory } = this.props;
    if (localStorage.getItem("auth_token")) {
      getOrderHistory();
    } else this.props.history.push("/sign-in");
  }

  componentWillReceiveProps(nextProps) {
    this.setState({ loading: false });
    if (nextProps.orderHistory.order) {
      this.setState({ orderHistory: nextProps.orderHistory.order });
    } else if (nextProps.orderHistoryError.response.status === 401) {
      this.props.history.push("/sign-in");
    }
  }

  render() {
    const { loading, orderHistory, orderHistoryError } = this.state;
    return (
      <Container style={{ padding: "1em 1em" }}>
        <Breadcrumb style={{ padding: "1em 0em" }}>
          <Breadcrumb.Section as={Link} to="/">
            Home
          </Breadcrumb.Section>
          <Breadcrumb.Divider>/</Breadcrumb.Divider>
          <Breadcrumb.Section as={Link} to="/menus">
            Menu
          </Breadcrumb.Section>
          <Breadcrumb.Divider>/</Breadcrumb.Divider>
          <Breadcrumb.Section active>Order History</Breadcrumb.Section>
        </Breadcrumb>
        {orderHistoryError && loading === false? (
          <Alert message="Error loading History..." type="negative" />
        ) : (
          ""
        )}
        {orderHistory.length < 1 && loading === false ? (
          <Alert message="You have no order History..." type="negative" />
        ) : (
          ""
        )}
        {loading ? (
          <Alert
            message="We are fetching your order history...."
            type="positive"
          />
        ) : (
          <Item.Group>
            {orderHistory.map((order, key) => (
              <OrderHistoryItem key={key} order={order} />
            ))}
          </Item.Group>
        )}
      </Container>
    );
  }
}

OrderHistory.protType = {
  getOrderHistory: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  orderHistory: state.menuReducer.orderHistory,
  orderHistoryError: state.menuReducer.orderHistoryError
});

export default connect(
  mapStateToProps,
  { getOrderHistory }
)(OrderHistory);
