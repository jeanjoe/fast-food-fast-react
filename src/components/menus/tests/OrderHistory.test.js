import React from "react";
import expect from "expect";
import Adapter from "enzyme-adapter-react-16";
import { shallow, configure } from "enzyme";
import { OrderHistory } from "../OrderHistory";
configure({ adapter: new Adapter() });

describe("Test article component", () => {
  const props = {
    getOrderHistory: jest.fn(),
    history: {
      push: jest.fn()
    }
  };
  const wrapper = shallow(<OrderHistory {...props} />);
  wrapper.setState({
    orderHistory: [],
  })
  it("Matches the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should handle componentDidMount on unauthenticated user", () => {
    wrapper.instance().componentDidMount();
    expect(props.history.push).toHaveBeenCalled();
  });
  it("should handle componentDidMount on authenticated user", () => {
    localStorage.setItem("auth_token", "sometoken");
    wrapper.instance().componentDidMount();
    expect(props.getOrderHistory).toHaveBeenCalled();
  });

  it("should handle componentWillReceiveProps", () => {
    const nextProps = {
      orderHistory: {
        order: [
          {
            id: 1,
            price: 500
          },
          {
            id: 2,
            price: 1000
          }
        ]
      },
      orderHistoryError: {}
    };
    wrapper.instance().componentWillReceiveProps(nextProps);
    expect(wrapper.state("orderHistory")).toEqual(nextProps.orderHistory.order);
  });
  it("should handle componentWillReceiveProps on Error", () => {
    const nextProps = {
      orderHistory: {
        order: null
      },
      orderHistoryError: {
        response: {
          status: 401
        }
      }
    };
    wrapper.instance().componentWillReceiveProps(nextProps);
    expect(props.history.push).toHaveBeenCalled();
  });
});
