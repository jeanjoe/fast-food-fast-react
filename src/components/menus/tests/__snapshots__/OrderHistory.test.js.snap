// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test article component Matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <OrderHistory
    getOrderHistory={[MockFunction]}
    history={
      Object {
        "push": [MockFunction] {
          "calls": Array [
            Array [
              "/sign-in",
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Breadcrumb
          style={
            Object {
              "padding": "1em 0em",
            }
          }
        >
          <BreadcrumbSection
            as={[Function]}
            to="/"
          >
            Home
          </BreadcrumbSection>
          <BreadcrumbDivider>
            /
          </BreadcrumbDivider>
          <BreadcrumbSection
            as={[Function]}
            to="/menus"
          >
            Menu
          </BreadcrumbSection>
          <BreadcrumbDivider>
            /
          </BreadcrumbDivider>
          <BreadcrumbSection
            active={true}
          >
            Order History
          </BreadcrumbSection>
        </Breadcrumb>,
        "",
        <Alert
          message="You have no order History..."
          type="negative"
        />,
        <Alert
          message="We are fetching your order history...."
          type="positive"
        />,
      ],
      "style": Object {
        "padding": "1em 1em",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <BreadcrumbSection
              as={[Function]}
              to="/"
            >
              Home
            </BreadcrumbSection>,
            <BreadcrumbDivider>
              /
            </BreadcrumbDivider>,
            <BreadcrumbSection
              as={[Function]}
              to="/menus"
            >
              Menu
            </BreadcrumbSection>,
            <BreadcrumbDivider>
              /
            </BreadcrumbDivider>,
            <BreadcrumbSection
              active={true}
            >
              Order History
            </BreadcrumbSection>,
          ],
          "style": Object {
            "padding": "1em 0em",
          },
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "as": [Function],
              "children": "Home",
              "to": "/",
            },
            "ref": null,
            "rendered": "Home",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "/",
            },
            "ref": null,
            "rendered": "/",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "as": [Function],
              "children": "Menu",
              "to": "/menus",
            },
            "ref": null,
            "rendered": "Menu",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "/",
            },
            "ref": null,
            "rendered": "/",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "active": true,
              "children": "Order History",
            },
            "ref": null,
            "rendered": "Order History",
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "",
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "message": "You have no order History...",
          "type": "negative",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "message": "We are fetching your order history....",
          "type": "positive",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Breadcrumb
            style={
              Object {
                "padding": "1em 0em",
              }
            }
          >
            <BreadcrumbSection
              as={[Function]}
              to="/"
            >
              Home
            </BreadcrumbSection>
            <BreadcrumbDivider>
              /
            </BreadcrumbDivider>
            <BreadcrumbSection
              as={[Function]}
              to="/menus"
            >
              Menu
            </BreadcrumbSection>
            <BreadcrumbDivider>
              /
            </BreadcrumbDivider>
            <BreadcrumbSection
              active={true}
            >
              Order History
            </BreadcrumbSection>
          </Breadcrumb>,
          "",
          <Alert
            message="You have no order History..."
            type="negative"
          />,
          <Alert
            message="We are fetching your order history...."
            type="positive"
          />,
        ],
        "style": Object {
          "padding": "1em 1em",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <BreadcrumbSection
                as={[Function]}
                to="/"
              >
                Home
              </BreadcrumbSection>,
              <BreadcrumbDivider>
                /
              </BreadcrumbDivider>,
              <BreadcrumbSection
                as={[Function]}
                to="/menus"
              >
                Menu
              </BreadcrumbSection>,
              <BreadcrumbDivider>
                /
              </BreadcrumbDivider>,
              <BreadcrumbSection
                active={true}
              >
                Order History
              </BreadcrumbSection>,
            ],
            "style": Object {
              "padding": "1em 0em",
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "as": [Function],
                "children": "Home",
                "to": "/",
              },
              "ref": null,
              "rendered": "Home",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "/",
              },
              "ref": null,
              "rendered": "/",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "as": [Function],
                "children": "Menu",
                "to": "/menus",
              },
              "ref": null,
              "rendered": "Menu",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "/",
              },
              "ref": null,
              "rendered": "/",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "active": true,
                "children": "Order History",
              },
              "ref": null,
              "rendered": "Order History",
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "message": "You have no order History...",
            "type": "negative",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "message": "We are fetching your order history....",
            "type": "positive",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
